### PR TITLE
Separate MediaMenu from MusicListView

### PR DESCRIPTION
**Pull request by danrabbit**
This is necessary for eventually using different types of widgets to represent Media such as in #350 

I get this is kind of a big diff. Some of these changes can probably be done in smaller steps if necessary. Let me know :)

### CODE REVIEW

File: src/Widgets/MediaMenu.vala [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+        if (generic_list.parent_wrapper.library.support_playlists () == false) {
+            media_menu_new_playlist.visible = false;
+        }
+        foreach (var playlist in generic_list.parent_wrapper.library.get_playlists ()) {
+            // Don't include this playlist in the list of available options
+            if (playlist == generic_list.playlist)
+                continue;
+
+            if (playlist.read_only == true)
```
Comment (BaptisteGelez) [inline src/Widgets/MediaMenu.vala]: Maybe add braces here while we are at it? And this condition and the above may be combined.

File: src/Widgets/MediaMenu.vala [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+    public void popup_media_menu (Gee.Collection<Media> selection) {
+        var media_menu_new_playlist = new Gtk.MenuItem.with_label (_("New Playlist…"));
+        media_menu_new_playlist.activate.connect (media_menu_new_playlist_clicked);
+
+        var add_to_playlist_menu = new Gtk.Menu ();
+        add_to_playlist_menu.append (media_menu_new_playlist);
+
+        if (generic_list.parent_wrapper.library.support_playlists () == false) {
+            media_menu_new_playlist.visible = false;
```
Comment (BaptisteGelez) [inline src/Widgets/MediaMenu.vala]: Why not just `media_menu_new_playlist.visible = generic_list.parent_wrapper.library.support_playlists ()`?

File: src/Widgets/MediaMenu.vala [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        }
+        add_to_playlist_menu.show_all ();
+        add_to_playlist.submenu = add_to_playlist_menu;
+
+        // if all medias are downloaded already, desensitize.
+        // if half and half, change text to 'Download %external of %total'
+        int temporary_count = 0;
+        int total_count = 0;
+        foreach (var m in selection) {
+            if (m.isTemporary)
```
Comment (BaptisteGelez) [inline src/Widgets/MediaMenu.vala]: Braces here too maybe?

File: src/Widgets/MediaMenu.vala [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                temporary_count++;
+            total_count++;
+        }
+
+        if (temporary_count < 1) {
+            import_to_library.sensitive = false;
+        } else {
+            import_to_library.sensitive = true;
+            if (temporary_count != total_count)
```
Comment (BaptisteGelez) [inline src/Widgets/MediaMenu.vala]: Braces again.

File: src/Widgets/MediaMenu.vala [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+            } else if (set_rating != m.rating) {
+                set_rating = 0;
+                break;
+            }
+        }
+
+        rate_media.rating_value = set_rating;
+
+        //remove the previous "Other Actions" submenu and create a new one
+        var contractorSubMenu = new Gtk.Menu ();
```
Comment (BaptisteGelez) [inline src/Widgets/MediaMenu.vala]: Maybe rename this variable to `contractor_sub_menu`.

**Comment by danrabbit** [comment]
@BaptisteGelez Yep you're right a lot of that is copy/paste. Thanks for being thorough! Should be all fixed :)

**Pull request merged**